File: solutions/253.meeting-rooms-ii/meeting-rooms-ii.js
/**
 * Definition for an interval.
 * function Interval(start, end) {
 *     this.start = start;
 *     this.end = end;
 * }
 */
/**
 * @param {Interval[]} intervals
 * @return {number}
 */
const minMeetingRooms = intervals => {
  let start = [],
    end = []

  intervals.forEach(interval => {
    start.push(interval.start)
    end.push(interval.end)
  })

  start.sort((a, b) => a - b)

  end.sort((a, b) => a - b)

  let rooms = 0
  let endpoint = 0

  for (var i = 0; i < intervals.length; i++) {
    if (start[i] < end[endpoint]) {
      rooms++
    } else {
      endpoint++
    }
  }
  return rooms
}

File: solutions/253.meeting-rooms-ii/meeting-rooms-ii.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e
import heapq


class Solution:
    def minMeetingRooms(self, intervals):
        intervals.sort(key=lambda x: x.start)
        heap = []  # stores the end time of intervals
        for i in intervals:
            if heap and i.start >= heap[0]:
                # no need to allocate
                heapq.heappop(heap)
                heapq.heappush(heap, i.end)
            else:
                heapq.heappush(heap, i.end)

        return len(heap)
